File: app/src/main/java/com/petnbu/petnbu/api/WebService.java
package com.petnbu.petnbu.api;

import android.arch.lifecycle.LiveData;

import com.petnbu.petnbu.model.Comment;
import com.petnbu.petnbu.model.Feed;
import com.petnbu.petnbu.model.UserEntity;

import java.util.List;

public interface WebService {

    LiveData<ApiResponse<Feed>> createFeed(Feed feed);

    LiveData<ApiResponse<Feed>> updateFeed(Feed feed);

    LiveData<ApiResponse<List<Feed>>> getGlobalFeeds(long after, int limit);

    LiveData<ApiResponse<List<Feed>>> getGlobalFeeds(String afterFeedId, int limit);

    LiveData<ApiResponse<List<Feed>>> getUserFeed(String userId, long after, int limit);

    LiveData<ApiResponse<List<Feed>>> getUserFeed(String userId, String afterFeedId, int limit);

    LiveData<ApiResponse<Feed>> getFeed(String feedId);

    LiveData<ApiResponse<UserEntity>> createUser(UserEntity userEntity);

    LiveData<ApiResponse<Feed>> likeFeed(String userId , String feedId);

    LiveData<ApiResponse<Feed>> unLikeFeed(String userId , String feedId);

    LiveData<ApiResponse<UserEntity>> getUser(String userId);

    void updateUser(UserEntity userEntity, SuccessCallback<Void> callback);

    LiveData<ApiResponse<List<Comment>>> getFeedComments(String feedId, long after, int limit);

    LiveData<ApiResponse<List<Comment>>> getCommentsPaging(String feedId, String commentId, int limit);

    LiveData<ApiResponse<Comment>> createFeedComment(Comment comment, String feedId);

    LiveData<ApiResponse<Comment>> likeComment(String userId, String commentId);

    LiveData<ApiResponse<Comment>> unLikeComment(String userId, String commentId);

    LiveData<ApiResponse<Comment>> createReplyComment(Comment comment, String parentCommentId);

    LiveData<ApiResponse<List<Comment>>> getSubComments(String commentId, long after, int limit);

    LiveData<ApiResponse<List<Comment>>> getSubCommentsPaging(String commentId, String afterCommentId, int limit);
}

File: app/src/main/java/com/petnbu/petnbu/api/WebService.kt
package com.petnbu.petnbu.api

import android.arch.lifecycle.LiveData

import com.petnbu.petnbu.model.Comment
import com.petnbu.petnbu.model.Feed
import com.petnbu.petnbu.model.Notification
import com.petnbu.petnbu.model.UserEntity

interface WebService {

    fun createFeed(feed: Feed): LiveData<ApiResponse<Feed>>

    fun updateFeed(feed: Feed): LiveData<ApiResponse<Feed>>

    fun getGlobalFeeds(after: Long, limit: Int): LiveData<ApiResponse<List<Feed>>>

    fun getGlobalFeeds(afterFeedId: String, limit: Int): LiveData<ApiResponse<List<Feed>>>

    fun getUserFeed(userId: String, after: Long, limit: Int): LiveData<ApiResponse<List<Feed>>>

    fun getUserFeed(userId: String, afterFeedId: String, limit: Int): LiveData<ApiResponse<List<Feed>>>

    fun getFeed(feedId: String): LiveData<ApiResponse<Feed>>

    fun createUser(userEntity: UserEntity): LiveData<ApiResponse<UserEntity>>

    fun likeFeed(userId: String, feedId: String): LiveData<ApiResponse<Feed>>

    fun unLikeFeed(userId: String, feedId: String): LiveData<ApiResponse<Feed>>

    fun getUser(userId: String): LiveData<ApiResponse<UserEntity>>

    fun updateUser(userEntity: UserEntity, callback: SuccessCallback<Void>)

    fun getFeedComments(feedId: String, after: Long, limit: Int): LiveData<ApiResponse<List<Comment>>>

    fun getCommentsPaging(feedId: String, commentId: String, limit: Int): LiveData<ApiResponse<List<Comment>>>

    fun createFeedComment(comment: Comment, feedId: String): LiveData<ApiResponse<Comment>>

    fun likeComment(userId: String, commentId: String): LiveData<ApiResponse<Comment>>

    fun unLikeComment(userId: String, commentId: String): LiveData<ApiResponse<Comment>>

    fun likeSubComment(userId: String, subCommentId: String): LiveData<ApiResponse<Comment>>

    fun unLikeSubComment(userId: String, subCommentId: String): LiveData<ApiResponse<Comment>>

    fun createReplyComment(subComment: Comment, parentCommentId: String): LiveData<ApiResponse<Comment>>

    fun getSubComments(commentId: String, after: Long, limit: Int): LiveData<ApiResponse<List<Comment>>>

    fun getSubCommentsPaging(commentId: String, afterCommentId: String, limit: Int): LiveData<ApiResponse<List<Comment>>>

    fun getNotifications(userId: String, after: Long, limit: Int): LiveData<ApiResponse<List<Notification>>>

}


